google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/CanonicalExtensionHeadersSerializer.java

```diff
@@ -81,4 +67,55 @@ public StringBuilder serialize(Map<String, String> canonicalizedExtensionHeaders
     // Concatenate all custom headers
     return serializedHeaders;
   }
+
+  public StringBuilder serialize(Map<String, String> canonicalizedExtensionHeaders) {
+    return serialize(canonicalizedExtensionHeaders, false);
+  }
+
+  public StringBuilder serializeHeaderNames(
+      Map<String, String> canonicalizedExtensionHeaders, boolean isV4) {
+    StringBuilder serializedHeaders = new StringBuilder();
+
+    if (canonicalizedExtensionHeaders == null || canonicalizedExtensionHeaders.isEmpty()) {
+      return serializedHeaders;
+    }
+    Map<String, String> lowercaseHeaders = getLowercaseHeaders(canonicalizedExtensionHeaders, isV4);
+
+    List<String> sortedHeaderNames = new ArrayList<>(lowercaseHeaders.keySet());
+    Collections.sort(sortedHeaderNames);
+
+    for (String headerName : sortedHeaderNames) {
+      serializedHeaders.append(headerName).append(HEADER_NAME_SEPARATOR);
+    }
+
+    serializedHeaders.setLength(serializedHeaders.length() - 1); // remove trailing semicolon
+
+    return serializedHeaders;
+  }
+
+  public StringBuilder serializeHeaderNames(Map<String, String> canonicalizedExtentionHeaders) {
+    return serializeHeaderNames(canonicalizedExtentionHeaders, true);
+  }
+
+  private Map<String, String> getLowercaseHeaders(
+      Map<String, String> canonicalizedExtensionHeaders, boolean isV4) {
+    // Make all custom header names lowercase.
+    Map<String, String> lowercaseHeaders = new HashMap<>();
+    for (String headerName : new ArrayList<>(canonicalizedExtensionHeaders.keySet())) {
+
+      String lowercaseHeaderName = headerName.toLowerCase();
+
+      // If present, remove the x-goog-encryption-key and x-goog-encryption-key-sha256 headers.
+      if ("x-goog-encryption-key".equals(lowercaseHeaderName)
+          || "x-goog-encryption-key-sha256".equals(lowercaseHeaderName)
+          || (isV4 && "x-goog-encryption-algorithm".equals(lowercaseHeaderName))) {
```

Comment:
At the moment `x-goog-encryption-key` and `x-goog-encryption-key-sha256` should be removed for both v2 and v4. There's an open question the GCS team if this is the case moving forward. (no-op for now).

How does knowing v4 help here?

Reply:
"x-goog-encryption-key" and "x-goog-encryption-key-sha256" are removed for both v2 and v4 here. "x-goog-encyrption-algorithm" is only removed for v4 here (C# does this as well) which is why there's a check

Reply:
Linking you to an internal bug: 128647687. This should be clarified when we get a response there.